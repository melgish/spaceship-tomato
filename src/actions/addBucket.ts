'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddBucket, addBucketSchema } from '@/schemas/addBucket.schema'

/**
 * Add server action to a new bucket to the database.
 * @see ../schemas/addBucket.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addBucket(payload: AddBucket) {
  try {
    const data = await addBucketSchema.parseAsync(payload)
    const created = await prisma.bucket.create({
      data: { ...data, plantId: data.plantId || null },
    })
    revalidatePath('/buckets')
    return { ok: true, message: 'Bucket created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create bucket' } as const
  }
}
