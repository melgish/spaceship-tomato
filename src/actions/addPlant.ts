'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddPlant, addPlantSchema } from '@/schemas/addPlant.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addPlant.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addPlant(payload: AddPlant) {
  try {
    const data = await addPlantSchema.parseAsync(payload)
    const created = await prisma.plant.create({ data })
    revalidatePath('/plants')
    return { ok: true, message: 'Plant created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create plant' } as const
  }
}
