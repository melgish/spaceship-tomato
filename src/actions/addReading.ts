'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddReading, addReadingSchema } from '@/schemas/addReading.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addReading.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addReading(payload: AddReading) {
  try {
    const data = await addReadingSchema.parseAsync(payload)
    const created = await prisma.reading.create({ data })
    revalidatePath('/readings')
    return { ok: true, message: 'Reading created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create reading' } as const
  }
}
