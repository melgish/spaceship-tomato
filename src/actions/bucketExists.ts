'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that a bucket exists in the database.
 * @param id Id of bucket to search for.
 * @returns true if bucket exists, otherwise false.
 */
export async function bucketExists(id: number = 0) {
  return id > 0 && (await prisma.bucket.count({ where: { id } })) > 0
}
