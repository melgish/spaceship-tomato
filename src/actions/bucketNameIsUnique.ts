'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that the name of a bucket is unique.
 * @param name Name of the bucket to test.
 * @param id Optional id of bucket to exclude when editing an existing bucket.
 * @returns true if the name exists on another bucket, otherwise false.
 */
export async function bucketNameIsUnique(name: string, id: number = 0) {
  const where = { name, id: (id > 0 && { not: id }) || undefined }
  return (await prisma.bucket.count({ where })) === 0
}
