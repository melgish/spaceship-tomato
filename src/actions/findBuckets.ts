'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of buckets from the database.
 * @returns An array of buckets. The returned data will include plant info if
 * the bucket contains a plant.
 */
export async function findBuckets() {
  return await prisma.bucket.findMany({
    include: { plant: true },
    orderBy: [{ name: 'asc' }],
  })
}

export type Buckets = Awaited<ReturnType<typeof findBuckets>>
export type Bucket = Buckets[number]
