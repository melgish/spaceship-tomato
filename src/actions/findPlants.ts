'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of plants from the database.
 * @returns An array of plants.
 */
export async function findPlants() {
  return await prisma.plant.findMany({
    orderBy: [{ name: 'asc' }],
  })
}

export type Plants = Awaited<ReturnType<typeof findPlants>>
export type Plant = Plants[number]
