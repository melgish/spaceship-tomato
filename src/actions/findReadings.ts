'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all readings from the database.
 * @returns An array of readings.  Readings will include bucket and plant info.
 */
export async function findReadings() {
  return await prisma.reading.findMany({
    orderBy: [{ createdAt: 'desc' }],
    include: {
      bucket: {
        include: {
          plant: true,
        },
      },
    },
  })
}

export type Readings = Awaited<ReturnType<typeof findReadings>>
export type Reading = Readings[number]
