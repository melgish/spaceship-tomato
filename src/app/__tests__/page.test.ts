import { type Mock, beforeEach, expect, it, vi, describe } from 'vitest'
import { render } from '@testing-library/react'
import { findBuckets } from '@/actions/findBuckets'
import HomePage from '@/app/page'

vi.mock('@/actions/findBuckets')

beforeEach(() => {
  ;(findBuckets as Mock).mockResolvedValue([
    {
      id: 1,
      name: 'Test Bucket',
      plant: { id: 1, name: 'Test Plant', minPH: 6.1, maxPH: 7.0 },
      plantId: 1,
      readings: [],
    },
  ])
})

describe('when there is no data', () => {
  it('should render', async () => {
    const { container } = render(await HomePage())
    expect(container).toMatchSnapshot()
  })
})
