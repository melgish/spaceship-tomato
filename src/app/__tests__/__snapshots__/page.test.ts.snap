// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when there is no data > should render 1`] = `
<div>
  <section>
    <header>
      <div>
        Home
      </div>
    </header>
    <div
      class="flex flex-wrap justify-around gap-y-2"
    >
      <div>
        <div>
          Bucket: 
          Test Bucket
           
          Test Plant
        </div>
        <div>
          PH
        </div>
        <svg
          font-family="sans-serif"
          font-size="10"
          height="156"
          stroke="currentColor"
          text-anchor="middle"
          width="344"
        >
          <g
            transform="translate(32,12)"
          >
            <g
              class="x axis"
              transform="translate(0, 120)"
            >
              <path
                d="M0,6V0H300V6"
                fill="none"
              />
            </g>
            <g
              class="y axis"
              text-anchor="end"
            >
              <path
                d="M-6,120H0V0H-6"
                fill="none"
              />
              <g
                class="tick"
                transform="translate(0, 120)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  5.0
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 100)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  5.5
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 80)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  6.0
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 60)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  6.5
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 40)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  7.0
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 20)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  7.5
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 0)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  8.0
                </text>
              </g>
            </g>
            <g
              class="bands"
              opacity="0.3"
              stroke="none"
            >
              <path
                d="M0 96V76H300V96Z"
                fill="#EE0"
              />
              <path
                d="M0 20V40H300V20Z"
                fill="#EE0"
              />
              <path
                d="M0 76V40H300V76Z"
                fill="#0E0"
              />
            </g>
            <g
              class="data"
              fill="forestgreen"
              stroke="forestgreen"
            >
              <path
                fill="none"
                stroke-width="2"
              />
            </g>
            <g
              class="no-data"
            >
              <text
                stroke="none"
                x="150"
                y="60"
              >
                No Data To Display
              </text>
            </g>
          </g>
        </svg>
        <div>
          EC
        </div>
        <svg
          font-family="sans-serif"
          font-size="10"
          height="156"
          stroke="currentColor"
          text-anchor="middle"
          width="344"
        >
          <g
            transform="translate(32,12)"
          >
            <g
              class="x axis"
              transform="translate(0, 120)"
            >
              <path
                d="M0,6V0H300V6"
                fill="none"
              />
            </g>
            <g
              class="y axis"
              text-anchor="end"
            >
              <path
                d="M-6,120H0V0H-6"
                fill="none"
              />
              <g
                class="tick"
                transform="translate(0, 115.2)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  800
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 105.6)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  900
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 96)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1000
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 86.4)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1100
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 76.8)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1200
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 67.2)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1300
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 57.6)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1400
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 48)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1500
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 38.4)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1600
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 28.8)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1700
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 19.2)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1800
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 9.6)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1900
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 0)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  2000
                </text>
              </g>
            </g>
            <g
              class="data"
              fill="forestgreen"
              stroke="forestgreen"
            >
              <path
                fill="none"
                stroke-width="2"
              />
            </g>
            <g
              class="no-data"
            >
              <text
                stroke="none"
                x="150"
                y="60"
              >
                No Data To Display
              </text>
            </g>
          </g>
        </svg>
        <div>
          TDS
        </div>
        <svg
          font-family="sans-serif"
          font-size="10"
          height="156"
          stroke="currentColor"
          text-anchor="middle"
          width="344"
        >
          <g
            transform="translate(32,12)"
          >
            <g
              class="x axis"
              transform="translate(0, 120)"
            >
              <path
                d="M0,6V0H300V6"
                fill="none"
              />
            </g>
            <g
              class="y axis"
              text-anchor="end"
            >
              <path
                d="M-6,120H0V0H-6"
                fill="none"
              />
              <g
                class="tick"
                transform="translate(0, 120)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  500
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 108)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  550
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 96)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  600
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 84)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  650
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 72)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  700
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 60)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  750
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 48)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  800
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 36)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  850
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 24)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  900
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 12)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  950
                </text>
              </g>
              <g
                class="tick"
                transform="translate(0, 0)"
              >
                <line
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  stroke="none"
                  x="-9"
                >
                  1000
                </text>
              </g>
            </g>
            <g
              class="data"
              fill="forestgreen"
              stroke="forestgreen"
            >
              <path
                fill="none"
                stroke-width="2"
              />
            </g>
            <g
              class="no-data"
            >
              <text
                stroke="none"
                x="150"
                y="60"
              >
                No Data To Display
              </text>
            </g>
          </g>
        </svg>
      </div>
    </div>
  </section>
</div>
`;
